File: jooq-generator/src/main/java/org/example/PostgresService.java
package org.example;

import org.gradle.api.services.BuildService;
import org.postgresql.ds.PGSimpleDataSource;
import org.testcontainers.containers.PostgreSQLContainer;

import javax.sql.DataSource;

public abstract class PostgresService
        implements BuildService<Params>, AutoCloseable {

    private PostgreSQLContainer<?> postgreSQLContainer;

    public PostgresService() {
        final var user = getParameters().getUser().convention("postgres");
        final var password = getParameters().getPassword().convention("password");

        this.postgreSQLContainer = new PostgreSQLContainer<>("postgres:16.4")
                .withUrlParam("ssl", "false")
                .withUsername(user.get())
                .withPassword(password.get());
        postgreSQLContainer.start();
    }

    public DataSource dataSource() {
        final var dataSource = new PGSimpleDataSource();

        dataSource.setUrl(postgreSQLContainer.getJdbcUrl());
        dataSource.setUser(postgreSQLContainer.getUsername());
        dataSource.setPassword(postgreSQLContainer.getPassword());

        return dataSource;
    }

    public String jdbc() {
        return postgreSQLContainer.getJdbcUrl();
    }

    public String username() {
        return postgreSQLContainer.getUsername();
    }

    public String password() {
        return postgreSQLContainer.getPassword();
    }

    @Override
    public void close() {
        postgreSQLContainer.stop();
    }
}

File: jooq-generator/src/main/kotlin/com/github/lipinskipawel/PostgresService.kt
package com.github.lipinskipawel

import org.gradle.api.services.BuildService
import org.postgresql.ds.PGSimpleDataSource
import org.testcontainers.containers.PostgreSQLContainer
import javax.sql.DataSource

abstract class PostgresService
    : BuildService<Params>, AutoCloseable {

    private var postgreSQLContainer: PostgreSQLContainer<*>

    init {
        val user = parameters.user.convention("postgres")
        val password = parameters.password.convention("postgres")

        postgreSQLContainer = PostgreSQLContainer("postgres:16.4")
            .withUrlParam("ssl", "false")
            .withUsername(user.get())
            .withPassword(password.get())
        postgreSQLContainer.start()
    }

    fun dataSource(): DataSource {
        val dataSource = PGSimpleDataSource()

        dataSource.setURL(postgreSQLContainer.getJdbcUrl())
        dataSource.user = postgreSQLContainer.username
        dataSource.password = postgreSQLContainer.password

        return dataSource
    }

    fun jdbc(): String {
        return postgreSQLContainer.jdbcUrl
    }

    fun username(): String {
        return postgreSQLContainer.username
    }

    fun password(): String {
        return postgreSQLContainer.password
    }

    override fun close() {
        postgreSQLContainer.stop()
    }
}
